### PR TITLE
Refactor `RequestDoc` functions (`get_all_descendants` and  `get_owner_to_type_dict`) to improve performance

### PR DESCRIPTION
**Pull request by dav3r**
#  #

## 🗣 Description ##

This pull request refactors the organization hierarchy traversal logic in `RequestDoc` to improve efficiency and testability. It introduces a pure in-memory traversal method, optimizes database queries to fetch only necessary fields, and adds comprehensive unit and integration tests for hierarchy traversal. The version is bumped to 1.3.1.

**Hierarchy traversal and performance improvements:**

* Refactored `get_all_descendants` and `get_owner_to_type_dict` in `RequestDoc` to use a new `_descendants` static method that performs pure in-memory traversal, and a new `_load_org_info` helper that fetches only required fields from the database, reducing overhead and improving performance. The traversal now avoids duplicate visits, handles cycles, and skips missing references robustly.

**Testing enhancements:**

* Added a new test suite `test_request_hierarchy.py` with unit tests for the in-memory traversal logic and integration tests that exercise the hierarchy traversal against a real MongoDB instance, ensuring correct handling of edge cases (e.g., cycles, shared subtrees, retired orgs, stakeholder filtering).

**Version bump:**

* Updated the package version to 1.3.1 in both `cyhy/__init__.py` and `setup.py`. [[1]](diffhunk://#diff-de176a738d1d82530aa97a7ebd47536547937246911d745a5dcefd4ad59e7b2cL9-R9) [[2]](diffhunk://#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7L5-R5)

## 💭 Motivation and context ##

During recent runs of the weekly reporting process, I noticed that it was taking over 5 minutes to build the list of snapshots to be generated:

```console
2026-07-13 13:06:12,019 INFO - Building list of reports to generate...
2026-07-13 13:06:12,064 INFO - Building list of snapshots to generate...
2026-07-13 13:11:55,500 INFO - [Thread-1] Starting snapshot for: CYHY_ENTITY
```

Digging into this, I saw that [we call `RequestDoc.get_all_descendants` many times](https://github.com/cisagov/cyhy-reports/blob/b4243269d045cb1d7d1b0bcead8e615ec3780ed9/extras/create_snapshots_reports_scorecard.py#L277) and I suspected that function was the reason why this step was taking such a long time to complete.

The changes in this PR result in significant performance improvements when doing a test against Production (over 12,000 entities) from my local system:

```python
# Legacy code
In [1]: import time
   ...: start_time = time.clock()
   ...: len(db.RequestDoc.get_all_descendants("ROOT"))
   ...: end_time = time.clock()
   ...: execution_time = end_time - start_time
   ...: print "Execution time: %.6f seconds" % execution_time
   ...:
Execution time: 4.017325 seconds

# Code from this PR
In [1]: import time
   ...: start_time = time.clock()
   ...: len(db.RequestDoc.get_all_descendants("ROOT"))
   ...: end_time = time.clock()
   ...: execution_time = end_time - start_time
   ...: print "Execution time: %.6f seconds" % execution_time
Execution time: 0.112626 seconds
``` 

NOTE: AI was used to generate most of the code in this PR.  I attest that I understand the generated code and am comfortable answering questions about it.

## 🧪 Testing ##

The new set of unit and integration tests related to these changes that was added in this PR all passed.

Additionally, I spot checked a variety of entities to ensure that the descendant counts were unchanged when compared with the original code, for example:

```python
In [1]: len(db.RequestDoc.get_all_descendants("ROOT"))
Out[1]: 12401

In [2]: len(db.RequestDoc.get_all_descendants("ROOT", include_retired=True))
Out[2]: 12816

In [3]: len(db.RequestDoc.get_all_descendants("ROOT", stakeholders_only=True, include_retired=True))
Out[3]: 11327
```

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.
- [x] Bump major, minor, patch, pre-release, and/or build versions [as appropriate](https://semver.org/#semantic-versioning-specification-semver) via the `bump_version` script *if* this repository is versioned *and* the changes in this PR [warrant a version bump](https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api).

## ✅ Pre-merge checklist ##

- [x] Finalize version.

## ✅ Post-merge checklist ##

- [x] Create a release (necessary if and only if the version was bumped).
- [x] Deploy this change to Production.

### CODE REVIEW

File: cyhy/db/database.py [hunk 1]
```diff
@@ -1280,52 +1280,81 @@ def remove_children(self, child_ids):
...
+        # cyclic data and avoids re-walking shared subtrees when an org is a
+        # child of more than one parent.  Note: retired orgs are pruned along
+        # with their entire subtree, while the stakeholder filter only affects
+        # membership (their children are still traversed).
         descendants = set()
-        descendants = descendants.union(
-            get_descendants(org_info, owner, stakeholders_only, include_retired)
-        )
+        visited = set()
+        stack = list(org_info[owner].get("children", []))
```
[the remark this answers is not in the document]
Comment (dav3r) [reply cyhy/db/database.py]: Good catch, please see https://github.com/cisagov/cyhy-core/pull/158/commits/2f4309415afd37ea6544c9541f7f7cb373908d6f where those suggestions have been implemented.

File: cyhy/test/test_request_hierarchy.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        assert len(result) == len(set(result))  # no dupes
+
+    def test_cycle_terminates(self):
+        # A -> B -> A ; must not infinite loop.
+        org_info = make_org_info(
+            make_org("A", ["B"]),
+            make_org("B", ["A"]),
+        )
+        result = RequestDoc._descendants(org_info, "A")
+        assert sorted(result) == ["A", "B"]
```
[the remark this answers is not in the document]
Comment (dav3r) [reply cyhy/test/test_request_hierarchy.py]: Taken care of in https://github.com/cisagov/cyhy-core/pull/158/commits/f8d23792cfa49d5c61b2e986d48a443ab43b9431 (along with addition of a couple other unit tests).

**Comment by jsf9k** [comment]
@dav3r - How are the unit tests expected to be run?  `pip install -r requirements.txt` followed by `pytest` is not sufficient.  Even after I run `pip install pytest mock` I still get an error:
```
❯ pytest
==================================================================================== test session starts ====================================================================================platform linux2 -- Python 2.7.18, pytest-4.6.11, py-1.11.0, pluggy-0.13.1
rootdir: /home/jeremy_frasier/cisagov/cyhy-core
collected 185 items / 1 errors / 184 selected                                                                                                                                               

========================================================================================== ERRORS ===========================================================================================______________________________________________________________________ ERROR collecting cyhy/test/test_nmap_handler.py ______________________________________________________________________ImportError while importing test module '/home/jeremy_frasier/cisagov/cyhy-core/cyhy/test/test_nmap_handler.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
cyhy/test/test_nmap_handler.py:13: in <module>
    from cyhy_commander.nmap.nmap_handler import NmapContentHandler
E   ImportError: No module named cyhy_commander.nmap.nmap_handler
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!================================================================================== 1 error in 0.62 seconds =================================================================================
```

Even an attempt to run just the new tests fails:
```
❯ pytest cyhy/test/test_request_hierarchy.py
============================= test session starts ==============================
platform linux2 -- Python 2.7.18, pytest-4.6.11, py-1.11.0, pluggy-0.13.1
rootdir: /home/jeremy_frasier/cisagov/cyhy-core
collected 15 items

cyhy/test/test_request_hierarchy.py ...........EEEE                      [100%]

==================================== ERRORS ====================================
_ ERROR at setup of TestGetAllDescendantsIntegration.test_default_excludes_retired_subtree_and_dedupes _

    @pytest.fixture
    def database():
        # connection = MongoClient('mongodb://tester:tester@[::1]:27017/test_database2', 27017)
        # db = connection['test_database2']
>       db = pcsdb.db_from_config("testing")

cyhy/test/common_fixtures.py:10: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cyhy/db/database.py:77: in db_from_config
    config = Config(config_section, config_filename)
cyhy/core/config.py:26: in __init__
    self.__write_config()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cyhy.core.config.Config object at 0x7f768bdc2650>

    def __write_config(self):
        config = SafeConfigParser()
        config.add_section(PRODUCTION_SECTION)
        config.add_section(TESTING_SECTION)
        # config section None goes to [DEFAULT]
        config.set(None, DEFAULT_SECTION, TESTING_SECTION)
        config.set(PRODUCTION_SECTION, DATABASE_NAME, PRODUCTION_DATABASE_NAME)
        config.set(TESTING_SECTION, DATABASE_NAME, TESTING_DATABASE_NAME)
        config.set(None, DATABASE_URI, DEFAULT_DATABASE_URI)
        config.set(None, REPORT_KEY, "")
>       with open(self.config_filename, "wb") as config_file:
E       IOError: [Errno 2] No such file or directory: '/etc/cyhy/cyhy.conf'

cyhy/core/config.py:53: IOError
___ ERROR at setup of TestGetAllDescendantsIntegration.test_include_retired ____

    @pytest.fixture
    def database():
        # connection = MongoClient('mongodb://tester:tester@[::1]:27017/test_database2', 27017)
        # db = connection['test_database2']
>       db = pcsdb.db_from_config("testing")

cyhy/test/common_fixtures.py:10: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cyhy/db/database.py:77: in db_from_config
    config = Config(config_section, config_filename)
cyhy/core/config.py:26: in __init__
    self.__write_config()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cyhy.core.config.Config object at 0x7f768bdc2b90>

    def __write_config(self):
        config = SafeConfigParser()
        config.add_section(PRODUCTION_SECTION)
        config.add_section(TESTING_SECTION)
        # config section None goes to [DEFAULT]
        config.set(None, DEFAULT_SECTION, TESTING_SECTION)
        config.set(PRODUCTION_SECTION, DATABASE_NAME, PRODUCTION_DATABASE_NAME)
        config.set(TESTING_SECTION, DATABASE_NAME, TESTING_DATABASE_NAME)
        config.set(None, DATABASE_URI, DEFAULT_DATABASE_URI)
        config.set(None, REPORT_KEY, "")
>       with open(self.config_filename, "wb") as config_file:
E       IOError: [Errno 2] No such file or directory: '/etc/cyhy/cyhy.conf'

cyhy/core/config.py:53: IOError
__ ERROR at setup of TestGetAllDescendantsIntegration.test_stakeholders_only ___

    @pytest.fixture
    def database():
        # connection = MongoClient('mongodb://tester:tester@[::1]:27017/test_database2', 27017)
        # db = connection['test_database2']
>       db = pcsdb.db_from_config("testing")

cyhy/test/common_fixtures.py:10: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cyhy/db/database.py:77: in db_from_config
    config = Config(config_section, config_filename)
cyhy/core/config.py:26: in __init__
    self.__write_config()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cyhy.core.config.Config object at 0x7f768bdc2a10>

    def __write_config(self):
        config = SafeConfigParser()
        config.add_section(PRODUCTION_SECTION)
        config.add_section(TESTING_SECTION)
        # config section None goes to [DEFAULT]
        config.set(None, DEFAULT_SECTION, TESTING_SECTION)
        config.set(PRODUCTION_SECTION, DATABASE_NAME, PRODUCTION_DATABASE_NAME)
        config.set(TESTING_SECTION, DATABASE_NAME, TESTING_DATABASE_NAME)
        config.set(None, DATABASE_URI, DEFAULT_DATABASE_URI)
        config.set(None, REPORT_KEY, "")
>       with open(self.config_filename, "wb") as config_file:
E       IOError: [Errno 2] No such file or directory: '/etc/cyhy/cyhy.conf'

cyhy/core/config.py:53: IOError
_ ERROR at setup of TestGetAllDescendantsIntegration.test_unknown_owner_raises _

    @pytest.fixture
    def database():
        # connection = MongoClient('mongodb://tester:tester@[::1]:27017/test_database2', 27017)
        # db = connection['test_database2']
>       db = pcsdb.db_from_config("testing")

cyhy/test/common_fixtures.py:10: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cyhy/db/database.py:77: in db_from_config
    config = Config(config_section, config_filename)
cyhy/core/config.py:26: in __init__
    self.__write_config()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cyhy.core.config.Config object at 0x7f768bdb35d0>

    def __write_config(self):
        config = SafeConfigParser()
        config.add_section(PRODUCTION_SECTION)
        config.add_section(TESTING_SECTION)
        # config section None goes to [DEFAULT]
        config.set(None, DEFAULT_SECTION, TESTING_SECTION)
        config.set(PRODUCTION_SECTION, DATABASE_NAME, PRODUCTION_DATABASE_NAME)
        config.set(TESTING_SECTION, DATABASE_NAME, TESTING_DATABASE_NAME)
        config.set(None, DATABASE_URI, DEFAULT_DATABASE_URI)
        config.set(None, REPORT_KEY, "")
>       with open(self.config_filename, "wb") as config_file:
E       IOError: [Errno 2] No such file or directory: '/etc/cyhy/cyhy.conf'

cyhy/core/config.py:53: IOError
====================== 11 passed, 4 error in 0.52 seconds ======================
```

**Comment by dav3r** [comment]
Thanks @jsf9k for opening #159 in response to [your comment above](https://github.com/cisagov/cyhy-core/pull/158#issuecomment-5025277275).

**Review by jsf9k — APPROVED** [review]
